### PR TITLE
SDI-318 async resource unit tests

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+  @MockBean
+  private lateinit var hmppsQueueService: HmppsQueueService
+
+  @MockBean
+  private lateinit var hmppsAsyncQueueService: HmppsAsyncQueueService
+
+  @Nested
+  inner class RetryDlq {
+    @Test
+    fun `should call the service for a sync queue client`() = runBlocking<Unit> {
```
Comment (mikehalmamoj) [inline hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt:40]: The first part of the magic for mocking suspend functions...

File: hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        .expectStatus().isOk
+        .expectBody()
+        .jsonPath("$.messagesFoundCount").isEqualTo(2)
+        .jsonPath("$.messages.length()").isEqualTo(1)
+
+      verify(hmppsQueueService).retryDlqMessages(check { it.hmppsQueue === hmppsQueue })
+    }
+
+    @Test
+    fun `should call the service for an async queue client`() = runBlocking<Unit> {
```
Comment (mikehalmamoj) [inline hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt:59]: The first piece of magic to mock suspend functions - gives us a CoroutineScope

File: hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+      val hmppsAsyncQueue = mock<HmppsAsyncQueue>()
+      whenever(hmppsQueueService.findByDlqName("some dlq name"))
+        .thenReturn(null)
+      whenever(hmppsAsyncQueueService.findByDlqName("some dlq name"))
+        .thenReturn(hmppsAsyncQueue)
+      whenever(hmppsAsyncQueueService.retryDlqMessages(any())).doSuspendableAnswer {
+        withContext(Dispatchers.Default) { RetryDlqResult(2, listOf(Message.builder().body("some body"))) }
+      }
```
Comment (mikehalmamoj) [inline hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt:67]: Suspendable functions can then be mocked with `doSuspendableAnswer`.

File: hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+      webTestClient.put()
+        .uri("/queue-admin/retry-dlq/some dlq name")
+        .exchange()
+        .expectStatus().isOk
+        .expectBody()
+        .jsonPath("$.messagesFoundCount").isEqualTo(2)
+        .jsonPath("$.messages.length()").isEqualTo(1)
+
+      verify(hmppsQueueService, never()).retryDlqMessages(any())
+      verify(hmppsAsyncQueueService).retryDlqMessages(RetryAsyncDlqRequest(hmppsAsyncQueue))
```
Comment (mikehalmamoj) [inline hmpps-sqs-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueResourceTest.kt:78]: And we can now verify the suspend function call as normal.

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**